cybozu/tcp.hpp

```diff
@@ -149,21 +154,65 @@ class tcp_socket: public resource {
     //
     // @return `true` if this socket is valid, `false` otherwise.
     bool sendv_close(const iovec* iov, int iovcnt) {
-        if( iovcnt >= MAX_IOVCNT )
-            throw std::logic_error("<tcp_socket::sendv> too many iov.");
-        lock_guard g(m_lock);
-        if( ! _sendv(iov, iovcnt, g) )
-            return false;
-        m_shutdown = true;
-        if( empty() ) {
-            _flush();
+        return with_fd([=](int fd) -> bool {
+            if( iovcnt >= MAX_IOVCNT )
+                throw std::logic_error("<tcp_socket::sendv> too many iov.");
+            lock_guard g(m_lock);
+            if( ! _sendv(fd, iov, iovcnt, g) )
+                return false;
+            m_shutdown = true;
+            if( empty() ) {
+                _flush(fd);
+                return false;
+            }
             g.unlock();
-            invalidate_and_close();
+            m_cond_write.notify_all();
             return true;
+        });
+    }
+
+    enum class recv_result {
+        OK,    // Received some data.
+        AGAIN, // No data was available.
+        RESET, // Connection reset by peer.
+        NONE,  // Peer half-closed the connection.
+    };
+
+    // Receive data from the socket.
+    // @buf           A buffer to store received data.
+    // @max_recvsize  Maximum size of data to be received.
+    //
+    // This function receives data from the socket.  Since this uses `with_fd`
+    // internally, the reactor thread should not call this, or it may be blocked.
```

Comment:
I think it is basically safe to call `with_fd` from the reactor thread.
If you call `with_fd` with write lock taken, it will deadlock, but write lock is only taken by `try_close` and `~resource`, and `with_fd` is not called from these method.

Reply:
you're right. will update the comment.